### ISSUE TITLE
BUG: Assertion Failed in `bcachefs: libbcachefs/btree_iter.c:1760: bch2_btree_path_peek_slot`

### ISSUE THREAD

**Issue by firestack** [koverstreet/bcachefs-tools#71]
I've been hitting the following bug when running `fsck` since commit `2b8c1bb0910534e8687ea3e5abf6d8bbba758247`

twice since running fsck I've hit a what seems to be early exit from fsck because of this bugcheck
```
❯ sudo bcachefs fsck /dev/sdb /dev/sdd /dev/sde /dev/sda /dev/sdc /dev/nvme0n1 -v -p
bcachefs: bch2_fs_open()
bcachefs: bch2_read_super()
bcachefs: bch2_read_super() ret 0
bcachefs: bch2_read_super()
bcachefs: bch2_read_super() ret 0
bcachefs: bch2_read_super()
bcachefs: bch2_read_super() ret 0
bcachefs: bch2_read_super()
bcachefs: bch2_read_super() ret 0
bcachefs: bch2_read_super()
bcachefs: bch2_read_super() ret 0
bcachefs: bch2_read_super()
bcachefs: bch2_read_super() ret 0
bcachefs: bch2_fs_alloc()
bcachefs: bch2_fs_journal_init()
bcachefs: bch2_fs_journal_init() ret 0
bcachefs: bch2_fs_btree_cache_init()
bcachefs: bch2_fs_btree_cache_init() ret 0
bcachefs: bch2_fs_encryption_init()
bcachefs: bch2_fs_encryption_init() ret 0
bcachefs: __bch2_fs_compress_init()
bcachefs: __bch2_fs_compress_init() ret 0
bcachefs: bch2_dev_alloc()
bcachefs: bch2_dev_alloc() ret 0
bcachefs: bch2_dev_alloc()
bcachefs: bch2_dev_alloc() ret 0
bcachefs: bch2_dev_alloc()
bcachefs: bch2_dev_alloc() ret 0
bcachefs: bch2_dev_alloc()
bcachefs: bch2_dev_alloc() ret 0
bcachefs: bch2_dev_alloc()
bcachefs: bch2_dev_alloc() ret 0
bcachefs: bch2_dev_alloc()
bcachefs: bch2_dev_alloc() ret 0
bcachefs: bch2_fs_alloc() ret 0
recovering from clean shutdown, journal seq 389740
journal read done, 0 keys in 1 entries, seq 389741
starting alloc read
alloc read done
starting stripes_read
stripes_read done
starting mark and sweep
mark and sweep done
starting journal replay
journal replay done
starting fsck
checking extents
checking dirents
going read-write
bcachefs: libbcachefs/btree_iter.c:1760: bch2_btree_path_peek_slot: Assertion `!(path->uptodate != BTREE_ITER_UPTODATE)' failed.
fish: Job 1, 'sudo bcachefs fsck /dev/sdb /de…' terminated by signal SIGABRT (Abort)
```

should I expect data loss? what should I provide to help debugging or fix this? ( I didn't see anything in the site or repo about debugging issues or other expected info with debug reports. I'm sorry if I've missed something)

**Comment by koverstreet** [comment]
No, you shouldn't expect data loss. I may have already fixed this kernel side; I'll update -tools to the latest bcachefs code. Once I do that, retest and if it still happens, run bcachefs fsck under gdb and get me a backtrace.

**Comment by firestack** [comment]
I do not have any debugging modes turned on in my kernel build or bcachefs tools build as far as I'm aware, and I'm on a nixos machine
```Linux etoix 5.13.0 #1-NixOS``` :| I'm compiling the kernel straight from the bcachefs repository with bcachefs as a module because default settings seemed to work. 
I can attempt to turn extra debugging on but I do not know where to start,

I've recompiled the kernel already and am preped to run fsck with gdb _I think_, 

Ideally there would be dwarf debugging symbols built alongside via the dwarffs crate but I've not worked with that yet and may have to start.

**Comment by koverstreet** [comment]
I already updated bcachefs-tools, and it does look like a bug I already fixed - make sure you've pulled the latest -tools and rebuild that!

**Comment by firestack** [comment]
Done and done! 👍🏻

**Comment by koverstreet** [comment]
Did it work?

**Comment by firestack** [comment]
Oh, still trying to ensure  I have the right flags or anything else needed for gdb before first try, I'm doing a quick change to ensure all the versions are correct, since I'm on NixOS. I'll post once I run the command and the full output

I could submit a PR adding the flake I've been building if you'd like

**Comment by firestack** [comment]
I think everything is good and fixed by 08802949e2601a1334cd1b600a1c3bec10f5cdb3 
thank you!
```
❯ sudo gdb --args bcachefs fsck /dev/sdb /dev/sdd /dev/sde /dev/sda /dev/sdc /dev/nvme0n1 -v -p
GNU gdb (GDB) 10.2
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-unknown-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from bcachefs...
(No debugging symbols found in bcachefs)
(gdb) start
Temporary breakpoint 1 at 0x404d40
Starting program: /nix/store/kmrkhl2f18zjynz2lh96jzhndxf1dxza-system-path/bin/bcachefs fsck /dev/sdb /dev/sdd /dev/sde /dev/sda /dev/sdc /dev/nvme0n1 -v -p
warning: File "/nix/store/jsp3h3wpzc842j0rz61m5ly71ak6qgdn-glibc-2.32-54/lib/libthread_db-1.0.so" auto-loading has been declined by your `auto-load safe-path' set to "$debugdir:$datadir/auto-load:/nix/store/0kiykyrnrpfhmjwxwx89kxr20hmf5304-gcc-10.3.0-lib".
To enable execution of this file add
	add-auto-load-safe-path /nix/store/jsp3h3wpzc842j0rz61m5ly71ak6qgdn-glibc-2.32-54/lib/libthread_db-1.0.so
line to your configuration file "/root/.gdbinit".
To completely disable this security protection add
	set auto-load safe-path /
line to your configuration file "/root/.gdbinit".
For more information about this security protection see the
"Auto-loading safe path" section in the GDB manual.  E.g., run from the shell:
	info "(gdb)Auto-loading safe path"
warning: Unable to find libthread_db matching inferior's thread library, thread debugging will not be available.
[New LWP 50991]
[New LWP 50992]
[New LWP 50993]
[New LWP 50994]
[New LWP 50995]
[New LWP 50996]
[New LWP 50997]

Thread 1 "bcachefs" hit Temporary breakpoint 1, 0x0000000000404d40 in main ()

(gdb) c
Continuing.
bcachefs: bch2_fs_open()
bcachefs: bch2_read_super()
bcachefs: bch2_read_super() ret 0
bcachefs: bch2_read_super()
bcachefs: bch2_read_super() ret 0
bcachefs: bch2_read_super()
bcachefs: bch2_read_super() ret 0
bcachefs: bch2_read_super()
bcachefs: bch2_read_super() ret 0
bcachefs: bch2_read_super()
bcachefs: bch2_read_super() ret 0
bcachefs: bch2_read_super()
bcachefs: bch2_read_super() ret 0
bcachefs: bch2_fs_alloc()
[New LWP 51488]
[New LWP 51489]
[New LWP 51490]
[New LWP 51491]
bcachefs: bch2_fs_journal_init()
bcachefs: bch2_fs_journal_init() ret 0
bcachefs: bch2_fs_btree_cache_init()
bcachefs: bch2_fs_btree_cache_init() ret 0
[New LWP 51492]
bcachefs: bch2_fs_encryption_init()
bcachefs: bch2_fs_encryption_init() ret 0
bcachefs: __bch2_fs_compress_init()
bcachefs: __bch2_fs_compress_init() ret 0
bcachefs: bch2_dev_alloc()
[New LWP 51493]
bcachefs: bch2_dev_alloc() ret 0
bcachefs: bch2_dev_alloc()
[New LWP 51494]
bcachefs: bch2_dev_alloc() ret 0
bcachefs: bch2_dev_alloc()
[New LWP 51495]
bcachefs: bch2_dev_alloc() ret 0
bcachefs: bch2_dev_alloc()
[New LWP 51496]
bcachefs: bch2_dev_alloc() ret 0
bcachefs: bch2_dev_alloc()
[New LWP 51497]
bcachefs: bch2_dev_alloc() ret 0
bcachefs: bch2_dev_alloc()
[New LWP 51498]
bcachefs: bch2_dev_alloc() ret 0
bcachefs: bch2_fs_alloc() ret 0
recovering from clean shutdown, journal seq 391795
journal read done, 0 keys in 1 entries, seq 391796
[New LWP 51542]
[New LWP 51543]
starting alloc read
alloc read done
starting stripes_read
stripes_read done
starting mark and sweep
mark and sweep done
starting journal replay
journal replay done
starting fsck
checking extents
checking dirents
going read-write
checking xattrs
checking root directory
checking inode nlinks
fsck done
ret 0
[New LWP 51544]
[New LWP 51545]
[New LWP 51546]
mounted with opts: metadata_replicas=2,data_replicas=2,metadata_target=ssd,foreground_target=ssd,background_target=hdd,promote_target=cache,noshard_inode_numbers,degraded,verbose,fsck,fix_errors
bcachefs: bch2_fs_open() ret 0
shutting down
[LWP 51546 exited]
[LWP 51545 exited]
[LWP 51544 exited]
flushing journal and stopping allocators
flushing journal and stopping allocators complete
[LWP 51493 exited]
[LWP 51495 exited]
[LWP 51494 exited]
[LWP 51497 exited]
[LWP 51496 exited]
[LWP 51498 exited]
[LWP 51542 exited]
marking filesystem clean
shutdown complete
[LWP 51492 exited]
[LWP 51491 exited]
[LWP 51490 exited]
[LWP 51489 exited]
[LWP 51488 exited]
[LWP 50997 exited]
[LWP 50996 exited]
[LWP 50995 exited]
[LWP 50994 exited]
[LWP 50993 exited]
[LWP 50992 exited]
[LWP 50991 exited]
[LWP 51543 exited]
[Inferior 1 (process 50987) exited normally]
(gdb)
```

**Comment by koverstreet** [comment]
Wonderful!
